jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/PerMessageDeflateExtension.java

```diff
@@ -438,22 +503,65 @@ private boolean inflate(Callback callback) throws DataFormatException
                         continue;
                     }
 
-                    finished = true;
+                    complete = true;
                     break;
                 }
             }
 
             Frame chunk = new Frame(_first ? _frame.getOpCode() : OpCode.CONTINUATION);
             chunk.setRsv1(false);
             chunk.setPayload(payload);
-            chunk.setFin(_frame.isFin() && finished);
+            chunk.setFin(_frame.isFin() && complete);
 
-            nextIncomingFrame(chunk, callback);
+            boolean succeedCallback = complete;
+            Callback frameCallback = _frameCallback;
+            WebSocketCoreSession coreSession = (WebSocketCoreSession)getCoreSession();
+            Callback payloadCallback = Callback.from(() ->
+            {
+                getBufferPool().release(payload);
+                if (succeedCallback)
+                    frameCallback.succeeded();
+            }, t ->
+            {
+                // The error needs to be forwarded to the CoreSession if callback is failed.
+                getBufferPool().release(payload);
+                failFlusher(t);
+                coreSession.processHandlerError(t, NOOP);
+            });
+            _demand.decrementAndGet();
+            nextIncomingFrame(chunk, payloadCallback);
```

Comment:
If this method throws, the `payload` is not released back into the buffer pool?
I think you should move the payload release in `onCompleteFailure()`.

Reply:
`nextIncomingFrame` is not supposed to throw that's why we give it a callback. When we eventually call application code we have a try/catch so if the application throws we fail the callback. So if `nextIncomingFrame` throws it is a bug.

If we release in `onCompleteFailure()` then we have to keep a full list of all the payload callbacks in the flusher and remove them when the callback is succeeded (the payloads might not be succeeded before next iteration of the flusher). I think this is not a good idea.